File: core/Services/UpdateChecker.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace ArtTimeTracker.Core.Services;

public class UpdateChecker
{
    private const string ReleasesUrl = "https://api.github.com/repos/Paulomat2025/TottisArschTracker/releases/latest";

    public record UpdateInfo(string Version, string DownloadUrl, string ReleaseUrl);

    public static async Task<UpdateInfo?> CheckForUpdateAsync(string currentVersion)
    {
        try
        {
            using var http = new HttpClient();
            http.DefaultRequestHeaders.Add("User-Agent", "TottisArschTracker");
            http.Timeout = TimeSpan.FromSeconds(10);

            var json = await http.GetStringAsync(ReleasesUrl);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            var tagName = root.GetProperty("tag_name").GetString()?.TrimStart('v') ?? "";
            var releaseUrl = root.GetProperty("html_url").GetString() ?? "";

            if (string.Compare(tagName, currentVersion, StringComparison.OrdinalIgnoreCase) <= 0)
                return null;

            var downloadUrl = releaseUrl;
            if (root.TryGetProperty("assets", out var assets))
            {
                foreach (var asset in assets.EnumerateArray())
                {
                    var name = asset.GetProperty("name").GetString() ?? "";
                    if (name.Contains("Windows", StringComparison.OrdinalIgnoreCase) && name.EndsWith(".zip"))
                    {
                        downloadUrl = asset.GetProperty("browser_download_url").GetString() ?? releaseUrl;
                        break;
                    }
                }
            }

            return new UpdateInfo(tagName, downloadUrl, releaseUrl);
        }
        catch
        {
            return null;
        }
    }
}

File: mac/ArtTimeTracker/Services/UpdateChecker.swift
import Foundation
import AppKit

@MainActor
class UpdateChecker {
    static let currentVersion = "1.0.0"

    static func checkForUpdate() {
        Task {
            guard let update = await fetchLatestRelease() else { return }

            let alert = NSAlert()
            alert.messageText = "Update verfügbar"
            alert.informativeText = "Version \(update.version) ist verfügbar!\n\nJetzt herunterladen?"
            alert.alertStyle = .informational
            alert.addButton(withTitle: "Herunterladen")
            alert.addButton(withTitle: "Später")

            if alert.runModal() == .alertFirstButtonReturn {
                if let url = URL(string: update.downloadUrl) {
                    NSWorkspace.shared.open(url)
                }
            }
        }
    }

    private static func fetchLatestRelease() async -> (version: String, downloadUrl: String)? {
        guard let url = URL(string: "https://api.github.com/repos/Paulomat2025/TottisArschTracker/releases/latest") else { return nil }

        var request = URLRequest(url: url)
        request.setValue("TottisArschTracker", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 10

        guard let (data, _) = try? await URLSession.shared.data(for: request),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tagName = json["tag_name"] as? String else { return nil }

        let version = tagName.hasPrefix("v") ? String(tagName.dropFirst()) : tagName

        guard version.compare(currentVersion, options: .numeric) == .orderedDescending else { return nil }

        let releaseUrl = json["html_url"] as? String ?? ""

        var downloadUrl = releaseUrl
        if let assets = json["assets"] as? [[String: Any]] {
            for asset in assets {
                let name = asset["name"] as? String ?? ""
                if name.lowercased().contains("macos") && name.hasSuffix(".dmg") {
                    downloadUrl = asset["browser_download_url"] as? String ?? releaseUrl
                    break
                }
            }
        }

        return (version, downloadUrl)
    }
}
